### PR TITLE
Zpi 238 backend state machine fix

### PR DESCRIPTION
**Pull request by JakubSamulski**
- state machine
- endpoint to check if reservation is possible i.e there are no accepted reservations in period /gameinstance/timeframes/avaliable
- fix so if a reservation starts at 29th nov and ends at 4th dec so the reservations avaliability responds whith
- unavaliable from 29th-30th and from 1st to 4th
- optional reservation comment
- avaliabilities only for accepted and rented instances
- state machine hopefully its on local branch
- more explicit error messages with state machine i.e as cannot change status from to as user
- check if reservations are sorted
- check if opinions are sorted
- add user opinion returns 201 with error
- return false in reservation details instead of throwing exception
- sometimes duration is 0 for 22/11 start and 24/11 end dates
- SHOULD WORK game instance avaliability cors
- check in game instance opinion and useropinion if resrevation has apropierate status

### CODE REVIEW

**Comment by pdrzazgaa** [comment]
state machine works well

File: backend/src/main/java/com/zpi/backend/game_instance/GameInstanceController.java [hunk 1]
```diff
@@ -151,12 +152,18 @@ public ResponseEntity<List<GameInstanceUnAvailabilityDTO>> getGameInstanceUnAvai
...
-    @GetMapping(value="/{uuid}/reservations")
+    @RequestMapping(method = GET,value="/{uuid}/reservations")
     @PreAuthorize("isAuthenticated()")
     public ResponseEntity<List<GameInstanceUnAvailabilityDTO>> getGameInstanceReservations(@PathVariable String uuid, @RequestParam String year, @RequestParam String month, Authentication authentication)
             throws GameInstanceDoesNotExistException, UserDoesNotExistException {
         return ResponseEntity.status(HttpStatus.OK)
                 .body(gameInstanceService.getGameInstanceAvailabilityReservation(authentication,uuid,year,month));
     }
 
+    @RequestMapping(method= GET,value = "/{uuid}/timeframes/avaliable")
```
Comment (pdrzazgaa) [inline backend/src/main/java/com/zpi/backend/game_instance/GameInstanceController.java:163]: No description for the endpoint

**Comment by pdrzazgaa** [comment]
Screenshot 2023-11-22 at 21 48 08
 Shouldn't return google id in the error description

File: backend/src/main/java/com/zpi/backend/reservations/Reservation.java [hunk 2]
```diff
@@ -59,12 +59,17 @@ public Reservation fromDTO(NewReservationDTO newReservationDTO, User renter, Gam
...
+        if(this.renterComment !=null)
+            this.renterComment = newReservationDTO.getRenterComment().orElse(null);
         this.timestamp = new Date(System.currentTimeMillis());
         this.duration = DateUtils.getDuration(startDate,endDate);
         this.reservationId = UUID.randomUUID().toString();
         return this;
     }
+
+    public int getDuration() {
+        return DateUtils.getDuration(startDate,endDate);
```
Comment (pdrzazgaa) [inline backend/src/main/java/com/zpi/backend/reservations/Reservation.java:71]: Duration does not take start date and end date in duration (always is 2 days less than it should be)

Comment (JakubSamulski) [reply backend/src/main/java/com/zpi/backend/reservations/Reservation.java:71]: *1 day less

**Review by pdrzazgaa — APPROVED** [review]

**Pull request merged**